Services/Models/InternalDeviceState.cs

```diff
@@ -0,0 +1,247 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Models
+{
+    public interface IInternalDeviceState
+    {
+        bool PropertyChanged { get; set; }
+        Dictionary<string, object> GetProperties();
+        bool HasProperty(string key);
+        object GetProperty(string key);
+        void SetProperty(string key, object value);
+        Dictionary<string, object> GetState();
+        void SetState(Dictionary<string, object> state);
+        bool HasStateValue(string key);
+        object GetStateValue(string key);
+        void SetStateValue(string key, object value);
+    }
+
+    /// <summary>
+    /// The InternalDeviceState manages the internal state of the simulated telemetry and
+    /// device properties. The state is unique to each device.
+    /// </summary>
+    public class InternalDeviceState : IInternalDeviceState
+    {
+        public const string CALC_TELEMETRY = "CalculateRandomizedTelemetry";
+
+        private ILogger log;
+
+        // TODO https://github.com/Azure/device-simulation-dotnet/issues/171
+        // consider using thread safe concurrent dictionary here in case multiple scripts
+        // try to modify the state at the same time. 
+
+        // The virtual state of the simulated device. The simulationState is
+        // periodically updated using an external script.
+        private Dictionary<string, object> simulationState;
+
+        // The virtual state of the simulated device properties. The reported properties,
+        // if changed, will be written to the IoT Hub.
```

Comment:
I would reduce the terms count, e.g. is "virtual state" something different from "internal state" and "simulation state" ? also, I think we don't need to mention the "if changed" bit here